src/server/frontend_wayland/idle_inhibit_v1.cpp

```diff
@@ -0,0 +1,173 @@
+
+    std::shared_ptr<IdleInhibitV1Ctx> const ctx;
+};
+
+class IdleInhibitorV1
+    : public wayland::IdleInhibitorV1
+{
+public:
+    IdleInhibitorV1(wl_resource* resource, std::shared_ptr<IdleInhibitV1Ctx> const& ctx);
+    ~IdleInhibitorV1();
+
+private:
+    // Not sure why, but removing this unused StateObserver stops the functionality of the IdleInhibit.
+    // It seems unlikely that it will ever actually be used.
+    struct StateObserver : ms::IdleStateObserver
+    {
+        StateObserver(IdleInhibitorV1* idle_inhibitor)
+                : idle_inhibitor{idle_inhibitor}
+        {
+        }
+
+        void idle() override;
+
+        void active() override;
+
+        IdleInhibitorV1* const idle_inhibitor;
+    };
+
+    std::shared_ptr<IdleInhibitV1Ctx> const ctx;
+    std::shared_ptr<StateObserver> const state_observer;
+
+    /// (Would be) called by the state observer
+    /// @{
+    void idle();
+
+    void active();
+    /// @}
+};
+}
+}
+
+auto mf::create_idle_inhibit_manager_v1(
+        wl_display* display,
+        std::shared_ptr<Executor> const& wayland_executor,
+        std::shared_ptr<scene::IdleHub> const& idle_hub)
+-> std::shared_ptr<mw::IdleInhibitManagerV1::Global>
+{
+    auto ctx = std::shared_ptr<IdleInhibitV1Ctx>{new IdleInhibitV1Ctx{wayland_executor, idle_hub}};
+    return std::make_shared<IdleInhibitManagerV1Global>(display, std::move(ctx));
+}
+
+mf::IdleInhibitManagerV1Global::IdleInhibitManagerV1Global(
+    wl_display *display,
+    std::shared_ptr<IdleInhibitV1Ctx> const& ctx)
+    : Global{display, Version<1>()},
+      ctx{ctx}
+{
+    // TODO - remove
+    mir::log_info("Client asking for IdleInhibitManagerV1Global");
+}
+
+void mf::IdleInhibitManagerV1Global::bind(wl_resource* new_resource)
+{
+    // TODO - remove
+    mir::log_info("Client calling IdleInhibitManagerV1Global::bind()");
+    new IdleInhibitManagerV1{new_resource, ctx};
+}
+
+mf::IdleInhibitManagerV1::IdleInhibitManagerV1(
+        wl_resource *resource,
+        std::shared_ptr<IdleInhibitV1Ctx> const& ctx)
+        : wayland::IdleInhibitManagerV1{resource, Version<1>()},
+          ctx{ctx}
+{
+    mir::log_info("Client asking for idle inhibit! (Manager)");
+}
+
+void mf::IdleInhibitManagerV1::create_inhibitor(struct wl_resource* id, struct wl_resource* surface)
+{
+    // TODO - remove log and snarky comment
+    mir::log_info("Client calling IdleInhibitManagerV1::create_inhibitor()");
+    (void)surface;
+
+    new IdleInhibitorV1{id, ctx};
+}
+
+mf::IdleInhibitorV1::IdleInhibitorV1(wl_resource *resource, std::shared_ptr<IdleInhibitV1Ctx> const& ctx)
+        : wayland::IdleInhibitorV1{resource, Version<1>()},
+          ctx{ctx},
+          state_observer{std::make_shared<StateObserver>(this)}
+{
+    mir::log_info("IdleInhibitorV1 created!");
+    ctx->idle_hub->inhibit_idle();
+    ctx->idle_hub->register_interest(state_observer, time::Duration(0));
```

Comment:
From the behavior I can gather from Firefox, it seems that the client doesn't call `create_inhibitor()` when the surface is created, but rather when idle inhibiting should take place (for example, when a YouTube video is played, it asks for an inhibitor object to be created, then later destroyed upon pausing or changing the tab.)

The client, however, remains unaware of the surface's focus state, so that is Mir's responsibility to track and handle.

Reply:
This code is not tracking the surface's focus state, it's tracking Mir's idle state.

Except it's not tracking Mir's idle state because the observer is doing nothing.

Except! it is critical to this PR functioning, because the 0-duration timeout is causing Mir to schedule the idle timeout for the current time, which apparently causes it to never be called, which means Mir never idles after this is registered and the alarm is reset.

This is obviously not the ideal way to get this behavior, and I believe the `state_observer` should be removed entirely.